Games: Game of Life

loki495 opened the pull request:
This is another version of Conway's Game Of Life. It can be run one "generation" at a time, or on a timer with a user-chosen delay, as well customizable number of rows/columns, stalling detection, clearing/randomizing the board, and the ability to toggle a cell on or off by clicking/dragging.

![DeepinScreenshot_select-area_20210428144554](https://user-images.githubusercontent.com/20133903/116482529-307d0000-a83a-11eb-87e0-768706256264.png)

linusg reviewed (changes requested):
Please remove the committed `CMakeCache.txt` and clean up the commit history - removal of logging code, "cleanups", import of broken code previously written all don't deserve their own commits. See https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md

sin-ack commented:
I think all of these commits should be a single commit, since you're adding the game itself.

sin-ack commented on Userland/Games/GameOfLife/Board.h:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    bool get_cell(size_t index) const
+    {
+        if (index > total_size() - 1)
+            return false;
+
+        return m_cells[index];
+    }
```
Earlier you did `get_cell(index) ? 1 : 0`. `bool` is already an integer with a value of `0` or `1`, so those are superfluous.

sin-ack commented on Userland/Games/GameOfLife/Board.cpp:
```diff
@@ -0,0 +1,86 @@
+    int row = index / columns();
+    int column = index % columns();
+
+    int top_left = get_cell(column - 1, row - 1) ? 1 : 0;
+    int top_mid = get_cell(column, row - 1) ? 1 : 0;
+    int top_right = get_cell(column + 1, row - 1) ? 1 : 0;
+    int left = get_cell(column - 1, row) ? 1 : 0;
+    int right = get_cell(column + 1, row) ? 1 : 0;
+    int bottom_left = get_cell(column - 1, row + 1) ? 1 : 0;
+    int bottom_mid = get_cell(column, row + 1) ? 1 : 0;
+    int bottom_right = get_cell(column + 1, row + 1) ? 1 : 0;
```
See my comment on `get_cell` about these.

sin-ack commented on Userland/Games/GameOfLife/main.cpp:
```diff
@@ -0,0 +1,169 @@
+{
+
+    auto app = GUI::Application::construct(argc, argv);
+    auto app_icon = GUI::Icon::default_icon("app-gameoflife");
```
You don't seem to have an icon for this. :^) Don't worry though, you can add one later.

bcoles replied:
If this is replacing the `Conway` application then the `conway` icons could be renamed to `gameoflife` and used here.

sin-ack commented on Userland/Games/GameOfLife/Board.cpp:
```diff
@@ -0,0 +1,86 @@
+        m_cells[i] = false;
+    }
+
+    srand(time(0));
```
Also consider using `AK::Random` over C functions.

loki495 replied:
I'm not exactly clear what should go on main.

sin-ack replied:
I meant the `srand`, since initializing global state is usually something that's done in `main`. :^)

loki495 replied:
Yeah, that makes more sense. Now I'm using AK::Random as you suggested

sin-ack commented on Userland/Games/GameOfLife/Board.cpp:
```diff
@@ -0,0 +1,86 @@
+void Board::randomize()
+{
+    for (size_t i = 0; i < total_size(); ++i)
+        set_cell(i, rand() % 2);
```
Consider using `AK::Random` instead.

linusg commented:
not sure why you requested a review - my previous comment hasn't been resolved at all... `CMakeCache.txt` is still there, last commit still says "There is also some cleanup of code, mostly from clang-format, and added copyright headers". Squash that into the appropriate previous commits.

loki495 commented:
> not sure why you requested a review - my previous comment hasn't been resolved at all... `CMakeCache.txt` is still there, last commit still says "There is also some cleanup of code, mostly from clang-format, and added copyright headers". Squash that into the appropriate previous commits.

You're right, that was premature and I didn't check I had pushed correctly. 

I've now deleted that file, addressed @sin-ack 's comments, and as per his suggestion, also combined all commits into just one since its just adding a self-contained unit.

awesomekling commented on Userland/Games/GameOfLife/Board.cpp:
```diff
@@ -0,0 +1,135 @@
+    : m_columns(columns)
+    , m_rows(rows)
+{
+
```
Please don't leave empty lines after opening a `{`.

awesomekling commented on Userland/Games/GameOfLife/Board.h:
```diff
@@ -0,0 +1,50 @@
+private:
+    bool calculate_next_value(size_t index);
+
+    size_t m_columns;
+    size_t m_rows;
```
Please always initialize primitive variables.

awesomekling commented on Userland/Games/GameOfLife/Board.h:
```diff
@@ -0,0 +1,50 @@
+    const Vector<bool>& cells() { return m_cells; }
+
+    void run_generation();
+    bool is_stalled() { return m_stalled; }
```
This should be `const` qualified.

awesomekling commented on Userland/Games/GameOfLife/Board.h:
```diff
@@ -0,0 +1,50 @@
+    bool get_cell(size_t column, size_t row) const;
+    bool get_cell(size_t index) const;
+
+    const Vector<bool>& cells() { return m_cells; }
```
This should be `const` qualified.

awesomekling commented on Userland/Games/GameOfLife/Board.h:
```diff
@@ -0,0 +1,50 @@
+    void set_cell(size_t column, size_t row, bool on);
+    void set_cell(size_t index, bool on);
+
+    bool get_cell(size_t column, size_t row) const;
+    bool get_cell(size_t index) const;
```
Please remove the superfluous `get_` prefix, per the [project coding style](https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#right-3)

awesomekling commented on Userland/Games/GameOfLife/Board.h:
```diff
@@ -0,0 +1,50 @@
+
+class Board {
+public:
+    Board(size_t columns, size_t rows);
```
The conventional order throughout this entire codebase is `row, column` instead of `column, row`

awesomekling commented on Userland/Games/GameOfLife/Board.h:
```diff
@@ -0,0 +1,50 @@
+    void randomize();
+
+private:
+    bool calculate_next_value(size_t index);
```
This should be `const` qualified.

awesomekling commented on Userland/Games/GameOfLife/BoardWidget.cpp:
```diff
@@ -0,0 +1,178 @@
+        }
+    }
+
+    delete m_board;
+    m_board = new Board(columns, rows);
```
Please don't use naked `new` and `delete`. Instead, use `make<Board>(...)` to create a new Board and store it in an `OwnPtr<Board>`.

awesomekling commented on Userland/Games/GameOfLife/BoardWidget.h:
```diff
@@ -0,0 +1,70 @@
+    virtual void paint_event(GUI::PaintEvent&) override;
+    virtual void mousemove_event(GUI::MouseEvent&) override;
+    virtual void mouseup_event(GUI::MouseEvent&) override;
+    virtual void mousedown_event(GUI::MouseEvent& event) override;
```
Please omit the `event` parameter name here.

awesomekling commented on Userland/Games/GameOfLife/BoardWidget.h:
```diff
@@ -0,0 +1,70 @@
+            m_last_cell_toggled = m_board->total_size();
+    }
+
+    size_t last_toggled() { return m_last_cell_toggled; }
+    bool is_toggling() { return m_toggling_cells; }
```
These should be `const` qualified.

awesomekling commented on Userland/Games/GameOfLife/BoardWidget.h:
```diff
@@ -0,0 +1,70 @@
+    void clear_cells() { m_board->clear(); }
+    void randomize_cells() { m_board->randomize(); }
+
+    int get_cell_size();
+    Gfx::IntSize get_board_offset();
+
+    size_t get_index_for_point(int x, int y);
```
These should be `const` qualified.

awesomekling commented on Userland/Games/GameOfLife/BoardWidget.h:
```diff
@@ -0,0 +1,70 @@
+
+    void run_generation();
+
+    int running_timer_interval() { return m_running_timer_interval; }
```
This should be `const` qualified.

awesomekling commented on Userland/Games/GameOfLife/main.cpp:
```diff
@@ -0,0 +1,169 @@
+
+int main(int argc, char** argv)
+{
+
```
Please don't leave empty lines after opening a `{`

awesomekling reviewed (changes requested):
If this is meant to supersede the `Conway` application, please also remove that one. :)

loki495 commented on Userland/Games/GameOfLife/BoardWidget.cpp:
```diff
@@ -0,0 +1,177 @@
+        }
+    }
+
+    m_board = make<Board>(rows, columns);
```
Is this a leak, as far as the previous Board m_cells?

linusg commented on Userland/Games/GameOfLife/BoardWidget.cpp:
```diff
@@ -0,0 +1,177 @@
+        on_running_state_change();
+}
+
+void BoardWidget::set_running(bool r)
+{
+    if (r == m_running)
+        return;
+
+    m_running = r;
```
Let's call this `running`:

```suggestion
void BoardWidget::set_running(bool running)
{
    if (running == m_running)
        return;

    m_running = running;
```

linusg commented on Userland/Games/GameOfLife/BoardWidget.cpp:
```diff
@@ -0,0 +1,177 @@
+    int width = rect().width() / m_board->columns();
+    int height = rect().height() / m_board->rows();
+
+    int size = width;
+
+    if (width > height) {
+        size = height;
+    }
+
+    return size;
```
This can be simplified as `return min(width, height);`

linusg commented on Userland/Games/GameOfLife/Board.cpp:
```diff
@@ -0,0 +1,134 @@
+
+void Board::toggle_cell(size_t index)
+{
+    if (index > total_size() - 1)
+        return;
+
```
Can we `VERIFY(index < total_size())` instead - I don't think we have to support the case where index is out of bounds?

Same for all of the below.

loki495 replied:
I agree for all the toggle/set methods, but for the cell getter it does have a use. In ```calculate_next_value``` where I get the values for neighboring cells, it is convenient that ```cell``` defaults to 0 when checking edge cells.

linusg replied:
Good point, works for me :+1:

linusg reviewed (commented):
This looks pretty cool! Got some more feedback, please also resolve the conflict in Conway by rebasing onto master :)

It looks like you haven't removed the Conway icons and app file (`Base/res/apps/Conway.af`) - please do that, and ideally add one for Game of Life instead.

Also, if you add a 32x32 icon (should be easy as the 16x16 icon can be upscaled), you can use `serenity_app()` instead of `serenity_bin()`, which will embed both icons in the executable, which in turn makes it show up in the system menu and various other places.

linusg commented on Base/res/apps/GameOfLife.af:
```diff
@@ -0,0 +1,4 @@
+[App]
+Name=GameOfLife
```
```suggestion
Name=Game Of Life
```

linusg commented on Userland/Games/GameOfLife/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+    main.cpp
+)
+
+serenity_bin(GameOfLife)
```
```suggestion
serenity_app(GameOfLife ICON app-gameoflife)
```

linusg reviewed (changes requested):
Please also ensure to insert a blank line after copyright headers, otherwise CI fails.
